### ISSUE TITLE
Coming from Mithril

### ISSUE THREAD

**Issue by leeoniya** [leeoniya/domvm#10]
Hey @mikegleasonjr,

Before you hack up Mithril, take a look at domvm. I had similar issues with Mithril as the ones you've been discussing. Personally, I wanted to gut and/or externalize the "magic" and make it opt-in, have the framework be both concise and no-surprises as to what the internals were doing. I eliminated global auto-redraw and made it an explicit thing and made all sub-components independently redraw-able. I took some of the great ideas of Mithril and domchanger and wrote a new animal. There is less structural enforcement, so you can compose apps as best fits you needs. I am still writing some demo apps to showcase its usage in complex situations, such as http://threaditjs.com/. Currently, it's 2x faster than Mithril on dbmonster and has nearly identical syntax for templates, so porting a small app for investigation shouldn't require massive changes.

Router is almost ready but still has a couple bugs w/ the History api to fix.

Mithril's auto-redraw observers are implemented fully externally in `/src/watch.js`. For example, this is how opt-in global auto-redraw works in domvm with props:

```js
function randInt(min, max) {
	return Math.floor(Math.random() * (max - min + 1)) + min;
}

// a view
function PersonView(vm, person) {
	return function() {
		return [".person",
			["strong", person.name],
			" ",
			["em", {onclick: randomizeAge}, person.age],
		]
	}
	
	function randomizeAge(e) {
		person.age(randInt(0,100));
	}
}

// create an observer
var w = domvm.watch(function(ev) {
	maryVm.redraw();
})

// some model/data
var mary = {
	name: "Mary",
	age: w.prop(25),      // use observer
};

var maryVm = domvm(PersonView, mary).mount(document.body);
```

I'd be happy to answer any questions about how to get specific things done if you're coming from Mithril.

I don't plan to support IE8, so if you need that, please ignore this message :)

cheers!

**Comment by lawrence-dol** [comment]
I am still lurking following development on this very closely, because I am deeply in favor of a component-oriented framework over MVx (or rather, one that makes component-first equal to MVx or any other pattern). I actually implement M(VC) in Mithril with the Mithril provided so-called "controller" object serving to hold instance-specific data for a given component, not methods. I personally feel that the Mithril component model did not work out as cleanly, nor as intuitively as I would have liked.

As well as that, the Mithril community seems very much at cross-purposes with each other, especially the drive to "modularize" a tiny library that should be kept small enough to not necessitate. Even that would have been fine if by modularize what was done was to cleanly separate Mithril into an independent core and independent separate modules on top of it; but instead the community has moved to a "tooled build" approach, which shuts me out from ever contributing because I've no interest in the added complexity or having Node.js on my system. 

That doesn't stop me using Mithril, but it does shut me out from contributing. I just don't believe in building code so complex that you need a dependency resolver -- all my code is designed to be cleanly separated and trivially assemble by a straight-forward minimize-and-concatenate tool which is a single small Java class built into my web server and which runs as an independent JAR for command-line use. I seek out and prefer zero-dependency libraries, because the alternative is almost always bloat-hell.

At this point, though, I am invested in Mithril, so paramount to my switching would be a clean path of transition, especially for views.

What do you think of the feasibility of DOMVM having an "m" implementation that would make a Mithril view directly usable by DOMVM? It would also need to map other "m" functions (or at least many of them) in terms of DOMVM. If that were there for views, I would seriously consider working in expanding it to full coverage for my projects. But it needs to be a realistic possibility. Something that would be enough for existing projects to make the transition. Until then, I am afraid that DOMVM may not offer enough for me to justify switching. 

I am lucky enough to have my major projects designed as independent tasks, so every one can have a different set of dependencies, but if DOMVM meets the goal and is committed to being the simplest thing that works and being built to allow added functions to be layered over it and documenting that as a guiding principle, then that's a project I can really get on board with. That and the documented requirement to use strict semicolons, because semicolon insertion is idiotic and dangerous and should be dropped from the language (not that that will ever happen now).

If I could make the switch I can justify contributing to DOMVM on work time, since it would be advancing all of my projects, no differently that enhancing my core JS library that they all use.

**Comment by leeoniya** [comment]
@lawrence-dol

> I personally feel that the Mithril component model did not work out as cleanly, nor as intuitively as I would have liked.

It took me a while to figure out what "controller" was, actually. And in fact I'm still not 100% clear, it's something between middleware and app/view state & api. I decided to look for greener pastures after the whole thread about clarifying/renaming it to `init()` or `constructor()` and the feared back-compat breakage.

> all my code is designed to be cleanly separated and trivially assemble by a straight-forward minimize-and-concatenate tool

I also prefer not to turn what should be a fun, powerful and fast language that can be written in notepad into enterprisey tool-burdened, IDE-required monstrosity stripped of its flexibility.

> What do you think of the feasibility of DOMVM having an "m" implementation that would make a Mithril view directly usable by DOMVM? It would also need to map other "m" functions (or at least many of them) in terms of DOMVM. If that were there for views, I would seriously consider working in expanding it to full coverage for my projects. But it needs to be a realistic possibility. Something that would be enough for existing projects to make the transition. Until then, I am afraid that DOMVM may not offer enough for me to justify switching.

If you want the m() syntax, this gets you 95% there:

```js
function m() {
	return Array.prototype.slice.call(arguments);
}
```

If you want observers and auto-redraw: `m.prop()`, `m.withAttr()`, `m.request()`, are implemented nearly identically to Mithril with slightly different names and externally to the core in `/src/watch.js`. Isomorphism is done. Routing is nearly ready (and is awesome, btw). I don't have an `m.trust()` equivalent yet. While I'd like you to switch, I cannot possibly provide a 100% transition layer for all Mithril's functionality, quirks-and-all. You will have to start using domvm for me to know what's missing or help you restructure your apps as you want.

> That and the documented requirement to use strict semicolons, because semicolon insertion is idiotic and dangerous and should be dropped from the language (not that that will ever happen now).

Please start by contributing feedback/bugs/questions/docs/wiki as a user rather than dissecting the internals. I've put a lot of thought into this lib and if you're having to hack around something, it's probably because there's insufficient docs for what goes where/when. The API, ergonomics, speed and correctness are #1; premature source code pedanticism: a distant #100.

> If I could make the switch I can justify contributing to DOMVM on work time, since it would be advancing all of my projects, no differently that enhancing my core JS library that they all use.

That would be great. But again, use it first before working on the lib itself. I guarantee that you would be re-implementing/hacking things unnecessarily if you just rely on the README.

**Comment by mikegleasonjr** [comment]
@leeoniya @lawrence-dol  It's amazing on how we think along the same lines. I too prefer a component/widget based framework. In fact, I created a utility function `components.mithril()` for my project that encapsulates a Mithril object and uniformises it's API:

```js
Widget1 = function(opts) {
    this.counter = m.prop(opts.seed);
};

components.mithril(Widget1, Base);

Widget1.prototype.render = function() {
    // when we have to render
    return m('div', {
        onclick: this.click.bind(this)
    }, this.counter());
};

Widget1.prototype.redrawn = function(element, initialized, context) {
    // after render, same as "config" in mithril
    // (not mandatory to define)
};

Widget1.prototype.unload = function() {
    // when widget is unloaded
    // (not mandatory to define)
};

Widget1.prototype.click = function(event) {
    event.preventDefault();
    m.redraw();
};
```

The widget is instantiated like a normal component:

```js
m.component(Widget1, {
    seed: 1
});
```

Like that, I don't feel the need to have a view separate from the controller. My Widgets are simple because I compose them with other widgets. This encourages re-use. I can derive from other widgets also. But I do more compositions than inheritance.

As DOMVM can redraw a subtree of the page independently, I would like to do `this.redraw()` instead of `m.redraw()` to re-render the widget only.

I like the fact that you created a tool to handle the redraw separately from the core. But my intuition tells me that I would write more code again instead of a simple `m.redraw()`. Even the simple example in the README contains more code than the one-liner `m.redraw()`. But again, the possibility is there for complex scenario.

So in the end I removed `m.deferred` from my fork of Mitrhil, removed the auto-redrawing except on route change and brought a good polyfill for Promises in userland. I modified `m.request()` to accept a nodejs-style callback and I did `m.request = Promise.denodeify(m.request);` in userland on page load. That way I don't have to change existing code calling `m.request`.

This feels so much better! I actually have less code.

I have already a couple of views so I feels it's a little bit late to port everything to DOMVM. But I will certainly abandon Mithril after this project to either build my own or use one like yours (unless there is a major turn of events).

I don't feel that Mithril is open enough to the community for me to participate a lot in it. Even if I would like to, I don't think the maintainers don't have much decision power. Plus, there seems to be a rewrite in development but no one knows what it contains and there's no document somewhere for proposals, input or simply to tell us what's in it. In the mean time, I believe they are all over the place when there are a lot of fundamental bugs currently in it. I hit 3 of them in my first week of using it. The API leaks the internal abstractions and sometimes there are more than a way of doing something. I never read that much the source code of a library to understand what it does. The API documentation is not enough. I learned things about the library while reading the blog posts. So to be fluent you have to read everything, and when you forgot something, it's hard to find it back.

Leo Horie is certainly more brilliant that I will ever be don't get me wrong. I am 36 and never achieved something like Mithril. But I have seen _a lot_ of frameworks and their ecosystem in my 15 years of doing websites and my feeling is that the opportunity will slip out of their hands by having too much of an academic approach and one-man-showrisim.

**Comment by lawrence-dol** [comment]
@leeoniya : I like what I see insofar as the direction and, contrary to some, totally agree with the above priorities. I would definitely use the library myself before beginning any serious contributions (35+ years of experience coding = not a fool).

I didn't know about `watch.js`; they are the three things I use the most. I need routing for one project. I am definitely not looking for a quirks and all layer; just enough to make switching small enough to swallow. Here I am feeling you out for the feasibility of doing that (by which I mean, me doing that and contributing it back if it proves successful). I don't *want* `m.` syntax any longer than it takes me to prove DOMVM meets my needs and, assuming it does, strip the `m.` layer from existing code.

I am not trying to be pedantic, just looking for someone who shares the same values I do as to code-quality, modularity/layering, and simplicity. I've found a direct inverse-correlation between liking/depending on auto-semicolon-insertion and the quality of code produced -- those people tend to eschew good design and principles in favor of simply typing less characters.

I took on Mithril as a dependency always with the underlying knowledge that if the project was abandoned or went off the rails I could fork it at the latest sensible point and personally take over maintenance -- if you've already effectively done the moral equivalent, and done it well, then I see no point in doing it again. 

After seeing multiple, awful, "be all that you can be" frameworks come and go, I've reached the conclusion that a small and solid basic SPA support library is far and away the best direction. I am looking for the best option to invest in, both in using it and contributing to it. I am also looking for someone with the wherewithal to strongly lead the project and not allow it to become a swing designed by a committee.

At this point I have a lot of Mithril-centric code; The question before me is, stick with Mithril or switch to DOMVM? It's not a decision I will make lightly; and the competence of the author/architect is a big part of that.

**Comment by leeoniya** [comment]
@mikegleasonjr

> I like the fact that you created a tool to handle the redraw separately from the core. But my intuition tells me that I would write more code again instead of a simple `m.redraw()`. Even the simple example in the README contains more code than the one-liner `m.redraw()`. But again, the possibility is there for complex scenario.

If you want global redraw there's nothing special required, you just call `vm.redraw()` on the top-level view - it's part of the core. Also in the core, you can call `vm.emit.redraw()` on any sub-view's `vm` and it will bubble a `.redraw()` request to the root view. If you want to be uniform, you can always use `vm.emit.redraw()`, including on the root view. So there's actually no more code than is required by Mithril's implicit, always-global redraw.

The only thing that `watch.js` provides (appropriately IMO) are auto-emitting mutation/promise obeservers (fetch, sync, prop) that reduce having to do the above "redraw boilerplate" everywhere. This is why they can be implemented completely outside the core.

> [Mithril's] API leaks the internal abstractions and sometimes there are more than a way of doing something.

Well, there are two sides to that coin. You can have "one way" to do things. However, without leaking some abstractions it is impossible to optimize certain cases or even have certain useful patterns. React is a great example of "one way" to do things, but is incredibly convention-bound and rigid. In huge teams this is good because everyone knows where things go, but you have to find crazy ways/gymnastics of accomplishing basic tasks or inhale another huge dependency. domvm also leaks some internals through the `vm` object but they are not part of the "normal" way of doing things. That being said, there are certainly a lot of ways of structuring code using domvm's official APIs - there''s enough flexibility to have an entire wiki of possible patterns. I started a short list already (probably not 100% accurate anymore since recent changes): https://github.com/leeoniya/domvm/wiki/View-Model-Patterns

> I never read that much the source code of a library to understand what it does. The API documentation is not enough. I learned things about the library while reading the blog posts. So to be fluent you have to read everything, and when you forgot something, it's hard to find it back.

The core API of domvm is incredibly thin specifically because it only aims to provide a stateful/composable view layer over your models/data. It does not seek to dictate your component or application structure, only to abstract the DOM in a pleasant, composable and performant way. You can pick and choose from many application structures given this flexibility. `watch.js` and `route.js` are completely external and are just one way of handling auto-redraw and routing/state transitions.

**Comment by mikegleasonjr** [comment]
@leeoniya Thanks for the reply, starred your project and will probably try it for my future ones. Now I have to do some real work and stop fighting with the frameworks. My major pain points are gone now that I forked.

When I was talking about multiple ways of doing things I had the `onunload` callback in mind. Specifically the one from the context passed to the config handler, and the one defined in the controller (which has a bug in it right now). They do the same thing.

Having flexibility OTOH is great like you said. I will take a look at the code :)

**Comment by leeoniya** [comment]
@lawrence-dol 

> I am also looking for someone with the wherewithal to strongly lead the project and not allow it to become a swing designed by a committee.

well, the preferable outcome is "refined by commitee" (tm) of like-minded people. it's difficult to develop something new in isolation from others' use-cases. inevitably, things get overlooked, edges are rough, idiomatic docs get put on the backburner.

the project core will remain focused, small and modular. but the wiki/docs of patterns, demos as well as polyfill recommendations will expand, perhaps by comittee. i have no intention of adding features and bloating the API to reduce code size for every use-case.

as for semicolons, i obviously use them and that won't change.

**Comment by leeoniya** [comment]
@vicnicius, read through this thread and feel free to ask any questions here. I'm not deeply familiar with Mithril's internals or knowing its recommended workarounds, therefore may not be able to explain the exact parallels in behavior you may be used to. Perhaps @lawrence-dol or @mikegleasonjr can chime in those cases. However, if you need to know how to get something specific done, I will be happy to give you a few options.

As mentioned above, the README is a good starting point but currently does not document all the features. Also, check out other issues tagged with "discuss" if interested, as well as the anemic wiki [1] :)

[1] https://github.com/leeoniya/domvm/wiki/View-Model-Patterns

**Comment by leeoniya** [comment]
I've published the router and split the lib into a few modules. The API has changed slightly.

`domvm()` is now `domvm.view()`

`vm.html()` is now `domvm.html(vm.node)`

**Comment by leeoniya** [comment]
@isiahmeadows

While you guys wait on a Mithril rewrite which addresses the multitude of issues with globals, attr handling and routing semantics, allow me again to encourage you to evaluate domvm, which was written exactly to solve these issues as well as the tight coupling, magic and redraw perf i stumbled over in Mithril.

domvm, in 11k min (< 5k gz) now has a decoupled lightweight SPA router, isomorphism as well as implementing all the good parts from Mithril (m.prop, m.request, m.withAttr) in a fully decoupled fashion. It is also 2x faster at global redraw, supports isolated sub-view redraw and even node-level patching.

I'm finishing up an implementation of threaditjs [1] which is helping me polish up the routing and error handling semantics. I will publish the code in the next few days.

I would appreciate your feedback since you are well aware of the pain points and can evaluate how effectively (and at what wiring cost) domvm addresses them.

I won't bug you any more, but feel free to chime in.

Thanks!

[1] http://threaditjs.com/

**Comment by lawrence-dol** [comment]
I may have the perfect project for this coming up in about a month. If that happens, I'll definitely start with DOMVM if it's ready for production projects.

**Comment by isiahmeadows** [comment]
@leeoniya My concern about this is that, from looking at the demos, that it starts to force *unnecessary* decoupling. The decoupling seems to require *more* state and complexity to do basic things like mounting and joining components. With Mithril, the problem was it was poorly written initially (a fundamental problem), and that's why Leo's working on a rewrite. It's really the only true fix.

**Comment by leeoniya** [comment]
@lawrence-dol it depends on your definition of "production ready" :) Different modules are at different stages.

The "core" vdom and view layer code is solid and pretty well tested [1]. There are probably some rough edges around maintaining focus/selectedIndex if elements must be detached from the DOM for reordering but those will be quick bugfixes with `document.activeElement`, `Node.tabindex`, etc. FWIW, I do personally use the view layer in production for some fairly logic-heavy and async fetch dependent widgets.

The isomorphism portion has also seen enough tweaking/bug fixing that I'm pretty confident about it. However, there are no explicit tests for it as it would basically require replicating the entire test suite for a post-hydrated tree. It's on the todo, though.

The mutation observers are probably 75%, but again it's hard to judge the status. They work, but do not have official tests (or docs yet).

The router/reverse-router is brand new, I got it working as envisioned (after several iterations) a few days ago in a way that cleanly accommodates the threaditjs demo. This is not yet production ready, however I do need more eyeballs on it for more diverse use-cases and requirements. There's no sub-routing, because it's not clear exactly how this should work yet since there is only one URL bar and one HTML5 history that must ideally somehow multiplex the combination of each view's/sub-view's state (similar to REST).

[1] http://leeoniya.github.io/domvm/test/

**Comment by leeoniya** [comment]
@isiahmeadows 

> My concern about this is that, from looking at the demos, that it starts to force unnecessary decoupling. The decoupling seems to require more state and complexity to do basic things like mounting and joining components.

Would you be able to provide a demonstration of what you mean in Mithril? I would be happy to show you the domvm version that requires the least code to accomplish the same thing.

The beauty of js (and closures in general) is that you *can* pass things around implicitly without requiring strict dependency injection everywhere. This is what I feel domvm gets right. It allows you to inject dependencies but if you prefer more "magic", you can just use whatever you choose to be in scope. The situation is much worse when you *require* the magic and globals/in-scope assumptions because then you have no way out without BC breakage. I think this is precisely why Mithril is in the situation that it's in. -- my $0.02

**Comment by leeoniya** [comment]
Ok, it was too tempting. I decided to put the code where my mouth is.

I took the link-rotator demo from mithril's home page [1] to see if I could write an adapter for domvm to use that code via copy paste. The result is here [2] see it in action [3]. The minified adapter size which accommodates the translation of features needed by this demo is 373 bytes. The full source for the adapter with comments is also in the repo.

I did have to make 1 concession since domvm currently does not block redraw during async/promise resolution, so I had to add `initialValue: []` to `m.request` to ensure that there was an array which could always be mapped over before `pages.json` came back.

Afterwards I decided to implement the same thing in domvm [4].

```js
function Demo() {
	this.pages = w.get("pages.json", w.prop([]));

	this.rotate = function() {
		this.pages().push(this.pages().shift());
	};
}

function DemoView(vm, demo) {
	return function() {
		return ["div",
			demo.pages().map(function(page) {
				return ["a", {href: page.url}, page.title];
			}),
			["button", {onclick: demo.rotate}, "Rotate links"]
		];
	};
}

// init a global watch w/auto-redraw handler
var w = domvm.watch().on(function() { demoVm.redraw(); });
// init state/model & create vm
var demoVm = domvm.view(DemoView, new Demo(), null, null, {watch: w});
// render to doc
demoVm.mount(document.getElementById("example"));
```

Do note that this is not the only way to structure this, it's just my preference given this demo. For example you could fully internalize the model, to make it more monolithic:

```js
function Demo(vm) {
	var pages = w.get("pages.json", w.prop([]));

	function rotate () {
		pages().push(pages().shift());
	};

	return function() {
		return ["div",
			pages().map(function(page) {
				return ["a", {href: page.url}, page.title];
			}),
			["button", {onclick: rotate}, "Rotate links"]
		];
	};
}

var w = domvm.watch().on(function() { demo.redraw(); });
var demo = domvm.view(Demo, null, null, null, {watch: w});
demo.mount(document.getElementById("example"));
```

Or more monolithic still:

```js
function Demo(vm) {
	var w = domvm.watch().on(function() { vm.redraw(); });
	vm.opts = {watch: w};

	var pages = w.get("pages.json", w.prop([]));

	function rotate () {
		pages().push(pages().shift());
	};

	return function() {
		return ["div",
			pages().map(function(page) {
				return ["a", {href: page.url}, page.title];
			}),
			["button", {onclick: rotate}, "Rotate links"]
		];
	};
}

domvm.view(Demo).mount(document.getElementById("example"));

```

The key takeaway is that in domvm, there's no distinction between "components", "modules", "models", "controllers". Everything is uniform. There are only models and composable view(s) of those models. You can retain state and expose methods from wherever is appropriate for your app. `selectAll()` might be a view method while `.fetchData()` might be a model method and you can call one from the other. Your views can *be* your monolithic models/controllers that just accept json data and your models can wrap one or several views and everything can have an API.

With Mithril, the "controller" is just a weirdly named model/state constructor that happens to be restricted to being paired with a single view. Why "Page" is commented as "model" I don't understand cause it's just a namespace in the example. I guess MVC terminology and I don't get along when applied to client-side frameworks :)

5:20am, time for bed :\

[1] http://mithril.js.org/
[2] https://github.com/leeoniya/domvm/tree/master/demos/mithril-adapter
[3] http://leeoniya.github.io/domvm/demos/mithril-adapter/
[4] https://leeoniya.github.io/domvm/demos/mithril-adapter/async-link-rotate.html

**Comment by leeoniya** [comment]
I've decided to change the async prop and fetch semantics to make their APIs more clear.

before:
```js
this.pages = w.get("pages.json", w.prop([]));
```
after:
```js
this.pages = w.prop([], w.get("pages.json"));
```

It is now blatantly obvious that `this.pages` is a prop. The first param is the initial value, and the second is any async promise (in this case it's the one returned by `w.get()` which is just a fetch abstraction w/auto-redraw injected into the chain). I don't think I will go Mithril's route of thennable props and async fetches which return props. I'm now happy with the clarity of the above changes which are practically self-documenting.

I've also added a `prop.freshen()` which will re-fetch (and auto-redraw in this case). The plan is to also add an optional lifetime to async props so that any redraw can auto-freshen them once they've gone stale.

Demos have been updated.

**Comment by lawrence-dol** [comment]
Maybe refresh instead of freshen?

-- 
Lawrence Dol
http://think.dolhub.com/ | http://tech.dolhub.com/
(Sent from my phone)

On Jan 30, 2016, at 11:56, Leon Sorokin <notifications@github.com> wrote:

I've decided to change the async prop and fetch semantics to make their APIs more clear.

before:

this.pages = w.get("pages.json", w.prop([]));
after:

this.pages = w.prop([], w.get("pages.json"));
It is now blatantly obvious that this.pages is a prop. The first param is the initial value, and the second is any async promise (in this case it's the one returned by w.get() which is just a fetch abstraction w/auto-redraw injected into the chain). I don't think I will go Mithril's route of thennable props and async fetches which return props. I'm now happy with the clarity of the above changes which are practically self-documenting.

I've also added a prop.freshen() which will re-fetch (and auto-redraw in this case). The plan is to also add an optional lifetime to async props so that any redraw can auto-freshen them once they've gone stale.

Demos have been updated.

**Comment by lawrence-dol** [comment]
How weird this tour in to showing a waiting indicator while the request is pending? Set a default of null to indicate to the view to show the throbber? But then it would need to revert to default on refresh. 

-- 
Lawrence Dol
http://think.dolhub.com/ | http://tech.dolhub.com/
(Sent from my phone)

On Jan 30, 2016, at 11:56, Leon Sorokin <notifications@github.com> wrote:

I've decided to change the async prop and fetch semantics to make their APIs more clear.

before:

this.pages = w.get("pages.json", w.prop([]));
after:

this.pages = w.prop([], w.get("pages.json"));
It is now blatantly obvious that this.pages is a prop. The first param is the initial value, and the second is any async promise (in this case it's the one returned by w.get() which is just a fetch abstraction w/auto-redraw injected into the chain). I don't think I will go Mithril's route of thennable props and async fetches which return props. I'm now happy with the clarity of the above changes which are practically self-documenting.

I've also added a prop.freshen() which will re-fetch (and auto-redraw in this case). The plan is to also add an optional lifetime to async props so that any redraw can auto-freshen them once they've gone stale.

Demos have been updated.

**Comment by lawrence-dol** [comment]
RE: the async redraw..

Blocking redraw while network requests are pending is worthless in anything more than a trivial demo. One can't freeze the UI while waiting, it's a terrible UX.

The first thing I did after starting with Mithril was to wrap m.request with my own function to set async=true for all requests unconditionally. 

-- 
Lawrence Dol
http://think.dolhub.com/ | http://tech.dolhub.com/
(Sent from my phone)

**Comment by leeoniya** [comment]
> Maybe refresh instead of freshen?

i'm not excited by freshen. but the "refresh" is too UI-burdened :(

When i think "refresh" i expect the UI to update. In the case of an async prop, it's not always a prop that affects the ui and it may not even be attached to any observer that invokes redraw. I want a term that implies a re-runnig the async func but not more. Maybe "update" is the most unassuming.

**Comment by lawrence-dol** [comment]
In the context, update seems like a better choice. 

-- 
Lawrence Dol
http://think.dolhub.com/ | http://tech.dolhub.com/
(Sent from my phone)

On Jan 31, 2016, at 19:31, Leon Sorokin <notifications@github.com> wrote:

Maybe refresh instead of freshen?

i'm not excited by freshen. but the "refresh" is too UI-burdened :(

When i think "refresh" i expect the UI to update. In the case of an async prop, it's not always a prop that affects the ui and it may not even be attached to any observer that invokes redraw. I want a term that implies a re-runnig the async func but not more. Maybe "update" is the most unassuming.

**Comment by leeoniya** [comment]
update() it is then.

> Blocking redraw while network requests are pending is worthless in anything more than a trivial demo.

Worthless it is not. Remember FOUC [1]? Async-loading interfaces are the javascript equivalent of this. I find Mithril's initial rendering behavior of the link rotator to be more pleasant than simply letting async run its course as domvm does. Personally, I'd much rather wait an extra 200ms and be shown the full ui instead of having everything bounce around and reflow as its asynchronous pieces load up. I always disable google's "Instant Search" because the flashing of the entire page as I type is deeply distracting and entirely unnecessary - it's showoff-ism that makes for a shit ui.

Unfortunately, the way to avoid this is by making it synchronous via:

1. Server side rendering (SSR/isomorphism).
2. Block initial render until all async promises in the tree settle.

Option 1 is the smoothest, but not everyone can or wants to run js on the back end.

For Option 2 to work, the promise implementation must support something like `.allSettled()`, which requires external promise status introspection which ES6 and many others don't offer. It can likely be polyfilled though I'm not sure how cleanly. Perhaps if it can be done in < 1k over vanilla ES6, it can be included as an optional add-on module.

The most likely outcome is that I'll add a per-vm or global option to disable redraw on `.mount()`s (if any async nodes exist) and provide a method that collects all promises from a vm branch into a flat array. This way you can decide when to invoke the first redraw based on their statuses if the promise implementation supports this.

I'm bummed that ES6 promises are so gutted and basically provide the lowest common A+ denominator. That means we're stuck with shitty options like this or always having to ship custom Promise flavors :(

[1] https://en.wikipedia.org/wiki/Flash_of_unstyled_content

**Comment by isiahmeadows** [comment]
@leeoniya FWIW, I'm also intermittently working on a framework on my own, which is also another take on vdom. Here's a short (and incomplete) list of what I plan on including:

1. Routers, etc. are all external.
2. There are two types of components: those that render and patch themselves, and those that render vdom trees. The framework itself takes care of the diffing for vdom components, and it emits a stream of bufferable updates to the components that render themselves. DOM components are just that: components.
3. There is 0 coupling to any browser API (but it uses them if they're available).
4. Components are somewhere between React components and Mithril components. They're either plain objects or plain classes, that implement a specific interface.
5. Subtree rendering is relatively easy, and there is a `shouldUpdate` like React's `shouldComponentUpdate` and Mithril's `{subtree: "retain"}`. I also have a `m.retain()` factory for similar purposes.
6. It's possible to partially apply components (not supporting children), and return new components. This is done to be more size- and speed-efficient by using `m.part(Component, {...attrs})`, instead of creating a closure.

Another item of note: `m(".foo")` would throw an error in the debug build. You would have to do something like `m(d.div(".foo"))` or `m(d(".foo"))`. It's slightly more verbose, but the framework knows nothing of the DOM. You have to use DOM components for this.

The benefit of this is easy integration with other libraries and frameworks (imagine using a jQuery plugin - you could just wrap that into its own component with ease) and portability (it's almost trivial to port this to, say, a native Android, iOS, or Windows app).

The factories return plain objects, but most users would probably prefer to use a factory. The objects are made to be computer-generated, and the types are denoted via a 32-bit mask internally, so it's intended for compilers, not for humans. The mask helps detect types quicker on the fly, and they're also used with existing nodes to add some metadata about those components, to also optimize the framework itself (the interfaces are assumed to never change, so I can use the mask to store relevant type metadata for fast, static dispatch).

I'm still mostly at the conceptual stage, but the API has been influenced by some implementation work I've done and pencil+paper programming (if you know what I mean :wink:).

**Comment by leeoniya** [comment]
@isiahmeadows neat. definitely poke me when there's something to look at.

with domvm, i'm not targeting ultra low-level vdom as the consumable API as there are already numerous, more verbose libraries that do this and are very fast: kivi, vidom, cito. domvm gets 0.8x-1.0x their speed while having enough sugar to be usable without needing a more pleasant api wrapped around it.

**Comment by isiahmeadows** [comment]
@leeoniya My focus isn't necessarily in size or being super lightweight. My recent work is going to show up soon on GitHub in the form of a test framework that still needs more tests. The framework's assertions are already self-hosted, testing the framework itself. I'm currently using Mocha, but that's subject to change. I've yet to come up with a solid name for it, though.

**Comment by isiahmeadows** [comment]
@leeoniya [Here's what I've got so far.](https://github.com/isiahmeadows/testiphile) CLI isn't ready, but everything else is stabilizing relatively quickly. :smile:

**Comment by leeoniya** [comment]
@isiahmeadows 

I quickly checked out techtonic. It's not immediately obvious to me what it offers over Mocha, Qunit or others. While this is usually the case for early-stage WIP projects, the readme should at least address this fundamental question in some manifesto; attention spans are short these days and time is money. I did this for domvm so that it's not immediately relegated to the "yet another vdom framework". FWIW, the fact that there's no browser port and the docs are essentially "read the source" already means I personally have 85% less motivation to play with it.

Anyhoos, I'd be more interested in your vdom adventures :)

**Comment by isiahmeadows** [comment]
It's still super early stage, hence the "read the source/etc", but I can
document the API better once I get a little more time. I've taken a
temporary hiatus to finish fixing my website's design.

**Comment by isiahmeadows** [comment]
@leeoniya [Docs exist now.](https://github.com/isiahmeadows/techtonic/tree/master/docs) Thought I'd let you know.

**Comment by leeoniya** [comment]
@StephanHoyer @JAForbes @EirikBirkeland @spacejack @ACXgit

> no fucking magic
 
Anecdotally, "magic", surprises, globals, perf and MVC-ness was specifically why I moved away from Mithril. The backwards-incompatible `rewrite` branch looks much better though as it appears to drop the "C" and boost perf by a large margin. Still calling it "Mithril" while being drastically different is strange but makes sense for marketing. Take a look at [domvm](https://github.com/leeoniya/domvm) and see if it strikes your fancy, the learning curve and API surface is also minimal. Else, back to your regularly scheduled programming. /pun

Recent discussion: https://www.reddit.com/r/javascript/comments/4ft52t/domvm_v11_a_fast_isomorphic_view_layer_router_in/

**Comment by isiahmeadows** [comment]
@leeoniya

Mind voicing this in lhorie/mithril.js#1026? I feel that's probably a
better location for that discussion.

I will note that you can only remove opinions so far before it starts
increasing boilerplate and becoming harder to learn. Common Lisp is another
good example of this - the object system was added similarly to Promises/A+
in JS, but the core is so simple it requires you to add extra boilerplate
to get some simpler things done.

**Comment by leeoniya** [comment]
@isiahmeadows 

> Mind voicing this in lhorie/mithril.js#1026? I feel that's probably a better location for that discussion.

i did not feel it was appropriate for me to knock Mithril in a thread praising it, out of respect. While I do want people to use domvm, I donct want to that hijacking asshole and do in fact respect Leo's work. I would not be pleased if someone was blatantly advertising their competing work in this repo either. I'd be happy to join if that would be welcomed.

The rewrite branch looks promising also, and removes a good amount of issues I had with Mithril when I decided to jump ship. What it will look like at 1.0 remains to be seen.

> I will note that you can only remove opinions so far before it starts increasing boilerplate and becoming harder to learn

domvm has just enough opinions in it to hit the sweet spot, IMO. It is easier to learn than Mithril, with less magic, fewer surprises/hacking and more flexibility. domvm trades Mithril's globals for some trivial wiring, BUT if you like globals you can do this wiring/shimming once and use it exactly like Mithril. If you look at the mithril-adapter demo in this repo, you can see how tiny the translation layer is. On top of that, domvm is vastly more powerful and faster. If you haven't played with it, i highly recommend you do to experience the difference in ergonomics.

Again, I don't think this sales pitch is appropriate for that thread. What do you think?

**Comment by JAForbes** [comment]
@leeoniya If you want to critique mithril, then by all means do so.  Criticism is healthy.  But using that as an opportunity for your own gain undermines the credibility of your criticism.

In that thread we are trying to focus on communicating concisely what mithril is, and why it is useful.  Adding your personal reasons for disliking mithril is potentially valuable information, but it isn't really on topic in that particular thread.

If you'd like to criticize mithril without advertising your own library, then I think creating a thread to do so is completely fine.

I think your library's design is interesting.  But I could not be more happy with mithril.  So please do not directly approach me to use your library again.  If I decide to use it, it will be based on my own research and at my own accord.  Contacting people without solicitation is an effective, but unethical marketing tactic.

**Comment by leeoniya** [comment]
> If you want to critique mithril, then by all means do so. Criticism is healthy. But using that as an opportunity for your own gain undermines the credibility of your criticism.

Well, yes and no. I've used and liked Mithril, I've run into issues, I've seen many people run into issues before and since and I wrote something that avoids all those issues. Mithril is what it is and criticizing it without offering the obvious alternative (even if it is somewhat self-serving, though nothing is being sold here) is a mute point. I'm not about to debate if my critique is correct or not, since the proof os already in the pudding. Whether that undermines my credibility is of course up to you to decide, but it's best to evaluate every library on its own merits.

> So please do not directly approach me to use your library again. If I decide to use it, it will be based on my own research and at my own accord. Contacting people without solicitation is an effective, but unethical marketing tactic.

I do apologize, and feel free to ignore the invite. This was simply a way to put the lib on people's radar.

**Comment by indus** [comment]
@leeoniya One question to your comparision of the "link-rotator demo" in this issue (https://github.com/leeoniya/domvm/issues/10#issuecomment-177151989): While Mithril keeps Model and View kind of seperated, your approach seems to combine them in an indivisible unit. While this is the only difference that I can see I wonder if it would still be possible to use a single Model in multiple independent Views? Does it just need one of the other proposed patterns?

**Comment by leeoniya** [comment]
@indus 

> your approach seems to combine them in an indivisible unit

I think perhaps the `ViewModel` teminology is to blame here. The "Model" in ViewModel is *not* your domain model, it's the constructor, state & api of the View (which I consider to be the live DOM).

The domain models (yours) are dep-injected into the views, like React, etc. You can absolutely have fully independent views of the same model or portions of it.

```js
function AgeView(ageVm, model) {
	return function() {
		return [".age", model.age];
	};
}

function NameView(nameVm, model) {
	return function() {
		return ["em.name", model.name];
	};
}

var person = {
	name: "John",
	age: 60,
}

var ageVm = domvm.view(AgeView, model);
var nameVm = domvm.view(NameView, model);
```

In https://github.com/leeoniya/domvm/issues/10#issuecomment-177151989, the Demo() is the domain model that's passed into the DemoView during init. They are otherwise not linked. Take a look through the README and see if it helps make sense of things.

**Comment by indus** [comment]
@leeoniya If you wouldn´t have answered already I would have deleted my comment now (after I had a closer look at the examples). Sorry for wasting your time ;-)

**Comment by yosbelms** [comment]
@indus, you may want take a look at [this](https://en.wikipedia.org/wiki/Model_View_ViewModel) article wich explain the pattern.
